File: Patterns/Iterator Pattern/Cpp/Waitress.cpp
#pragma once

#include <iostream>
#include "Iterator/Iterator.cpp"
#include "Menu/PancakeHouseMenu.cpp"
#include "Menu/DinerMenu.cpp"

using namespace std;

class Waitress {
private:
    PancakeHouseMenu* pancakeHouseMenu;
    DinerMenu* dinerMenu;

public:
    Waitress(PancakeHouseMenu* pancakeHouseMenu, DinerMenu* dinerMenu)
        : pancakeHouseMenu(pancakeHouseMenu), dinerMenu(dinerMenu) {}

    void printMenu() {
        Iterator* pancakeIterator = pancakeHouseMenu->createIterator();
        Iterator* dinerIterator = dinerMenu->createIterator();

        cout << "메뉴\n----\n아침 메뉴\n";
        printMenu(pancakeIterator);

        cout << "\n점심 메뉴\n";
        printMenu(dinerIterator);
    }

    void printMenu(Iterator* iterator) {
        while (iterator->hasNext()) {
            MenuItem* menuItem = iterator->next();
            cout << menuItem->getName() << ", ";
            cout << menuItem->getPrice() << " -- ";
            cout << menuItem->getDescription() << "\n";
        }
    }
};

File: Patterns/Iterator Pattern/CSharp/Waitress.cs
using System;

using Iterator;
using Menu;

public class Waitress
{
    private Menu.PancakeHouseMenu pancakeHouseMenu;
    private Menu.DinnerMenu dinnerMenu;

    public Waitress(Menu.PancakeHouseMenu pancakeHouseMenu, Menu.DinnerMenu dinnerMenu) 
    {
        this.pancakeHouseMenu = pancakeHouseMenu;
        this.dinnerMenu = dinnerMenu;
    }

    public void PrintMenu() 
    {
        Iterator.Iterator pancakeIterator = pancakeHouseMenu.CreateIterator();
        Iterator.Iterator dinnerIterator = dinnerMenu.CreateIterator();

        Console.WriteLine("메뉴\n----\n아침 메뉴");
        PrintMenu(pancakeIterator);

        Console.WriteLine("\n점심 메뉴");
        PrintMenu(dinnerIterator);
    }

    public void PrintMenu(Iterator.Iterator iterator)
    {
        while (iterator.HasNext())
        {
            Menu.MenuItem menuItem = iterator.Next();
            Console.WriteLine($"{menuItem.Name}, {menuItem.Price} -- {menuItem.Description}");
        }
    }
}